lib/src/main/scala/spinal/lib/tools/ModuleAnalyzer.scala

```diff
@@ -0,0 +1,176 @@
+package spinal.lib.tools
+
+import spinal.core._
+
+import scala.language._
+import scala.collection.mutable
+
+/**
+ * Module topology analyzer. It provides some methods that return the input or output pins,
+ * all sub-modules or sub-blackboxes, all clocks inside, and apply condition to the returning
+ * results.
+ * @param module the module being analyzed
+ */
+class ModuleAnalyzer(module: Module) {
+  /**
+    * Return all input port of the module
+    * @return set of base type
+    */
+  def allInputs: mutable.LinkedHashSet[BaseType] = module.getAllIo.asInstanceOf[mutable.LinkedHashSet[BaseType]].filter(_.isInputOrInOut)
+
+  /**
+    * Filter the input ports
+    * @param filter - the predicate to filter the input ports
+    * @return set of filtered input ports
+    */
+  def getInputs(filter: BaseType => Boolean): mutable.LinkedHashSet[BaseType] = allInputs.filter(filter)
+
+  /**
+    * Get all output ports
+    * @return set of output base type
+    */
+  def allOutputs: mutable.LinkedHashSet[BaseType] = module.getAllIo.asInstanceOf[mutable.LinkedHashSet[BaseType]].filter(_.isOutputOrInOut)
+
+  /**
+    * Filter the output ports
+    * @param filter - the predicate to filter the output ports
+    * @return set of filtered output ports
+    */
+  def getOutputs(filter: BaseType=> Boolean): mutable.LinkedHashSet[BaseType] = allOutputs.filter(filter)
```

Comment:
```suggestion
  def getOutputs(filter: BaseType => Boolean = _ => true): mutable.LinkedHashSet[BaseType] = allOutputs.filter(filter)
```

Reply:
Same as `allOutputs`.

Reply:
True, my issue is that `getOutputs` is documented "Filter the output ports" and not "Get the outputs" as we would expect. So if I want to get the output, I type `getOutputs`, I notice it misses a parameter.

What do you think about this renaming?

```scala
def outputs: ... = ...
def outputsWhere(cond: ... => Boolean)
```